### PR TITLE
Ability to get connection strings from external provider at runtime.

### PR DESCRIPTION
**Pull request by mzabolotko**

### CODE REVIEW

File: Sources/Contour/Configurator/IConnectionStringProvider.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Contour.Configurator
+{
+    /// <summary>
+    /// Provides a connection string per label.
+    /// Overrides a connection string from a configuration section.
```
Comment (semion-akimtsev) [inline Sources/Contour/Configurator/IConnectionStringProvider.cs]: This seems to be the responsibility of AppConfigConfigurator

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 2]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+            [Test]
+            public void should_load_connection_string_provider_if_present()
+            {
+                const string name = "name";
+                const string provider = "provider";
+                string Config = $@"<endpoints>
+                                       <endpoint name=""{name}"" connectionString="""" connectionStringProvider=""{provider}"" />
+                                   </endpoints>";
+
+                var resoverMock = new Mock<IDependencyResolver>();
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: resoLver

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 3]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+
+            [Test]
+            public void should_not_load_connection_string_provider_if_not_present()
+            {
+                const string name = "name";
+                string Config = $@"<endpoints>
+                                       <endpoint name=""{name}"" connectionString="""" />
+                                   </endpoints>";
+
+                var resoverMock = new Mock<IDependencyResolver>();
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: L

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 4]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+                    rm => rm.Resolve(
+                        It.IsAny<string>(),
+                        It.Is<Type>(type => type == typeof(IConnectionStringProvider))),
+                    Times.Never,
+                    "Should not use a dependency resolver to load the connection string provider implementation.");
+            }
+
+            [Test]
+            public void should_set_connection_string_for_endpoint_provider_if_present()
+            {
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Clarification needed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 5]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+                        </endpoint>
+                
+                    </endpoints>";
+
+                var connectionStringProviderMock = new Mock<IConnectionStringProvider>();
+                connectionStringProviderMock
+                    .Setup(cspm => cspm.GetConnectionString(It.Is<MessageLabel>(l => l.Name == Label)))
+                    .Returns(AnotherString);
+
+                var resoverMock = new Mock<IDependencyResolver>();
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: L

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 6]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+
+                var senderConfiguration = configuration.SenderConfigurations.First(sc => sc.Label.Equals(MessageLabel.From(Label)));
+
+                var senderOptions = (RabbitSenderOptions)senderConfiguration.Options;
+                senderOptions.GetConnectionString().Value.Should().Be(LabelString, "Should use a connection string from the outgoing label.");
+            }
+
+
+            [Test]
+            public void should_set_connection_string_to_incoming_label_from_provider_if_present()
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: should_set_connection_string_ON_incoming_label_from_provider_if_present

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 7]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+                var configuration = (BusConfiguration)configurator.Configure(Name, busConfiguration);
+
+                var receiverConfiguration = configuration.ReceiverConfigurations.First(sc => sc.Label.Equals(MessageLabel.From(Label)));
+
+                var receiverOptions = (RabbitReceiverOptions)receiverConfiguration.Options;
+                receiverOptions.GetConnectionString().Value.Should().Be(AnotherString, "Should use a connection string from the provider for incoming label.");
+            }
+
+            [Test]
+            public void should_override_connection_string_to_incoming_label_from_provider_if_present()
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: should_override_connection_string_ON_incoming_label_from_provider_if_present

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 8]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+                        </endpoint>
+                
+                    </endpoints>";
+
+                var connectionStringProviderMock = new Mock<IConnectionStringProvider>();
+                connectionStringProviderMock
+                    .Setup(cspm => cspm.GetConnectionString(It.Is<MessageLabel>(l => l.Name == Label)))
+                    .Returns(AnotherString);
+
+                var resoverMock = new Mock<IDependencyResolver>();
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: L

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 4 above)
Comment (semion-akimtsev) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: should_use_endpoint_connection_string_even_if_provider_is_present

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 9]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: L

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 10]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+                        </endpoint>
+                
+                    </endpoints>";
+
+                var connectionStringProviderMock = new Mock<IConnectionStringProvider>();
+                connectionStringProviderMock
+                    .Setup(cspm => cspm.GetConnectionString(It.Is<MessageLabel>(l => l.Name == Label)))
+                    .Returns((string)null);
+
+                var resoverMock = new Mock<IDependencyResolver>();
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: L

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 11]
```diff
@@ -1116,6 +1116,372 @@ public void should_set_connection_string_if_present()
...
+                        </endpoint>
+                
+                    </endpoints>";
+
+                var connectionStringProviderMock = new Mock<IConnectionStringProvider>();
+                connectionStringProviderMock
+                    .Setup(cspm => cspm.GetConnectionString(It.Is<MessageLabel>(l => l.Name == Label)))
+                    .Returns(AnotherString);
+
+                var resoverMock = new Mock<IDependencyResolver>();
```
Comment (semion-akimtsev) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: L

**Review by semion-akimtsev — CHANGES_REQUESTED** [review]
Please fix artifacts' naming.

File: Sources/Contour/Configurator/IConnectionStringProvider.cs (hunk 1 above)
Comment (mzabolotko) [reply Sources/Contour/Configurator/IConnectionStringProvider.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 2 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 3 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 4 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 5 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 6 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 7 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 8 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 9 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 10 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 11 above)
Comment (mzabolotko) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs]: Fixed

**Pull request merged**